tests/wpt/grouping_formatter.py

```diff
@@ -71,119 +155,56 @@ def suite_end(self, _):
-            output += make_subtests_failure(test_name, failures, stack)
-        return output
-
-    def test_end(self, data):
+    def test_end(self, data: dict) -> Optional[UnexpectedResult]:
         self.completed_tests += 1
         test_status = data["status"]
-        test_name = data["test"]
-        had_unexpected_test_result = "expected" in data
-        subtest_failures = self.subtest_failures.get(test_name, [])
-
+        test_path = data["test"]
         del self.running_tests[data['thread']]
 
+        had_unexpected_test_result = "expected" in data
+        subtest_failures = self.subtest_failures.pop(test_path, [])
         if not had_unexpected_test_result and not subtest_failures:
             self.expected[test_status] += 1
             return None
 
         # If the test crashed or timed out, we also include any process output,
         # because there is a good chance that the test produced a stack trace
         # or other error messages.
+        stack = data.get("stack", None)
         if test_status in ("CRASH", "TIMEOUT"):
-            stack = self.test_output[test_name] + data.get('stack', "")
-        else:
-            stack = data.get('stack', None)
+            stack = f"\n{stack}" if stack else ""
+            stack = f"{self.test_output[test_path]}{stack}"
+
+        result = UnexpectedResult(
+            test_path,
+            test_status,
+            data.get("expected", test_status),
+            data.get("message", ""),
+            data["time"],
+            stack,
+            subtest_failures
+        )
 
-        output = ""
         if had_unexpected_test_result:
-            self.test_failures.append(data)
-            self.unexpected_tests[test_status].append(data)
-            lines = self.get_lines_for_unexpected_result(
-                test_name,
-                test_status,
-                data.get('expected', None),
-                data.get('message', None),
-                stack)
-            output += self.wrap_and_indent_lines(lines, "  ")
-
+            self.unexpected_tests[result.actual].append(data)
         if subtest_failures:
-            self.tests_with_failing_subtests.append(test_name)
-            output += self.get_output_for_unexpected_subtests(test_name,
-                                                              subtest_failures)
-        self.unexpected_results.append(
-            UnexpectedResult(test_name, test_status, output))
-        return output
+            self.tests_with_failing_subtests.append(data)
 
-    def test_status(self, data):
-        if "expected" in data:
-            self.subtest_failures[data["test"]].append(data)
+        self.unexpected_results.append(result)
+        return result
+
+    def test_status(self, data: dict):
+        if "expected" not in data:
+            return
+        self.subtest_failures[data["test"]].append(UnexpectedSubtestResult(
+            data["test"],
+            data.get("subtest", ""),
```

Comment:
can this one be made a strict `data["subtest"]`? do some subtest data dicts lack one?

Reply:
Looking at the other formatters that are included with the base `wptrunner` code it seems that this assume the existence of`subtest`. I'll make that change.

Thanks for the review!